sw/cheri/checks/pinmux_checker.cc

```diff
@@ -0,0 +1,495 @@
+/**
+ * Copyright lowRISC contributors.
+ * Licensed under the Apache License, Version 2.0, see LICENSE for details.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "pinmux_checker.hh"
+#include "../common/timer-utils.hh"
+
+/**
+ * Sets the value of some GPIO output pin. This change will only be visible
+ * if the pin has its corresponding `output_enable` bit set to 1, so that the
+ * pin is in output mode.
+ */
+inline void set_gpio_output(Capability<volatile SonataGpioFull> gpio, GpioPin pin, bool value) {
+  const uint32_t bit_mask    = (1 << pin.bit);
+  const uint32_t write_value = ((value ? 1 : 0) << pin.bit);
+  const uint8_t instance     = static_cast<uint8_t>(pin.instance);
+  uint32_t output            = gpio->instances[instance].output;
+  output &= ~bit_mask;
+  output |= write_value;
+  gpio->instances[instance].output = output;
+}
+
+/**
+ * Sets the output enable value of some GPIO pin. Setting this to 1 places the
+ * pin in output mode, and to 0 places it in input mode.
+ */
+inline void set_gpio_output_enable(Capability<volatile SonataGpioFull> gpio, GpioPin pin, bool value) {
+  const uint32_t bit_mask    = (1 << pin.bit);
+  const uint32_t write_value = ((value ? 1 : 0) << pin.bit);
+  const uint8_t instance     = static_cast<uint8_t>(pin.instance);
+  uint32_t output_enable     = gpio->instances[instance].output_enable;
+  output_enable &= ~bit_mask;
+  output_enable |= write_value;
+  gpio->instances[instance].output_enable = output_enable;
+}
+
+/**
+ * Get the value of some GPIO input pin. This will only be visible if the pin
+ * has its corresponding `output_enable` bit set to 0, so that the pin is in
+ * input mode.
+ */
+inline bool get_gpio_input(Capability<volatile SonataGpioFull> gpio, GpioPin pin) {
+  const uint32_t bit_mask = (1 << pin.bit);
+  const uint8_t instance  = static_cast<uint8_t>(pin.instance);
+  return (gpio->instances[instance].input & bit_mask) > 0;
+}
+
+/**
+ * Tests that the given UART appears to be working as expected, by sending and
+ * receiving some data using the given UART block. To pass, this test expects
+ * the Tx and the Rx of the specified UART to be manually connected on the
+ * board, so that data that is written can be read back by the board.
+ *
+ * @param read_timeout_msec The timeout in milliseconds provided to read each
+ * individual byte sent over UART. Any timeout causes immediate failure.
+ * @param test_length The number of random bytes to try sending over UART.
+ *
+ * Returns true if the test passed, or false if it failed.
+ */
+static bool uart_send_receive_test(ds::xoroshiro::P32R8 &prng, UartPtr uart, uint32_t read_timeout_msec,
```

Comment:
This is also exists in the pinmux test. Would you mind making a common library for shared functions.

Reply:
Sure! This was just duplicated because I didn't want the PRs to be dependent upon each other, making a longer critical path for reviews. Now that the pinmux test PR is merged I'm happy to migrate this to a common library.